frame/support/src/traits/misc.rs

```diff
@@ -20,9 +20,279 @@
+	where
+		T: Default,
+	{
+		match self {
+			Ok(inner) => inner,
+			Err(e) => {
+				debug_assert!(false, "{}: {:?}", DEFENSIVE_OP_INTERNAL_ERROR, e);
+				frame_support::log::error!(
+					target: "runtime",
+					"{}: {:?}",
+					DEFENSIVE_OP_PUBLIC_ERROR,
+					e
+				);
+				Default::default()
+			},
+		}
+	}
+}
+
+impl<T, E: sp_std::fmt::Debug> DefensiveResult<T, E> for Result<T, E> {
+	fn defensive_map_err<F, O: FnOnce(E) -> F>(self, o: O) -> Result<T, F> {
+		self.map_err(|e| {
+			debug_assert!(false, "{}: {:?}", DEFENSIVE_OP_INTERNAL_ERROR, e);
+			frame_support::log::error!(
+				target: "runtime",
+				"{}: {:?}",
+				DEFENSIVE_OP_PUBLIC_ERROR,
+				e
+			);
+			o(e)
+		})
+	}
+
+	fn defensive_map_or_else<U, F: FnOnce(T) -> U, D: FnOnce(E) -> U>(self, default: D, f: F) -> U {
+		self.map_or_else(
+			|e| {
+				debug_assert!(false, "{}: {:?}", DEFENSIVE_OP_INTERNAL_ERROR, e);
+				frame_support::log::error!(
+					target: "runtime",
+					"{}: {:?}",
+					DEFENSIVE_OP_PUBLIC_ERROR,
+					e
+				);
+				default(e)
+			},
+			f,
+		)
+	}
+
+	fn defensive_ok(self) -> Option<T> {
+		match self {
+			Ok(inner) => Some(inner),
+			Err(e) => {
+				debug_assert!(false, "{}: {:?}", DEFENSIVE_OP_INTERNAL_ERROR, e);
+				frame_support::log::error!(
+					target: "runtime",
+					"{}: {:?}",
+					DEFENSIVE_OP_PUBLIC_ERROR,
+					e
+				);
+				None
+			},
+		}
+	}
+}
+
+impl<T> DefensiveOption<T> for Option<T> {
+	fn defensive_map_or_else<U, F: FnOnce(T) -> U, D: FnOnce() -> U>(self, f: F, default: D) -> U {
+		self.map_or_else(
+			|| {
+				debug_assert!(false, "{}", DEFENSIVE_OP_INTERNAL_ERROR);
+				frame_support::log::error!(
+					target: "runtime",
+					"{}",
+					DEFENSIVE_OP_PUBLIC_ERROR,
+				);
+				default()
+			},
+			f,
+		)
+	}
+
+	fn defensive_ok_or_else<E, F: FnOnce() -> E>(self, err: F) -> Result<T, E> {
+		self.ok_or_else(|| {
+			debug_assert!(false, "{}", DEFENSIVE_OP_INTERNAL_ERROR);
+			frame_support::log::error!(
+				target: "runtime",
+				"{}",
+				DEFENSIVE_OP_PUBLIC_ERROR,
+			);
+			err()
+		})
+	}
+}
+
+/// A variant of [`DefensiveUnwrap`] with the same rationale, for the arithmetic operations where in
+/// case an infallible operation fails, it saturates.
+pub trait DefensiveSaturating {
+	/// Add `self` and `other` defensively.
+	fn defensive_saturating_add(self, other: Self) -> Self;
+	/// Subtract `other` from `self` defensively.
+	fn defensive_saturating_sub(self, other: Self) -> Self;
+	/// Multiply `self` and `other` defensively.
+	fn defensive_saturating_mul(self, other: Self) -> Self;
+}
+
+// NOTE: A bit unfortunate, since T has to be bound by all the traits needed. Could make it
+// `DefensiveSaturating<T>` to mitigate, but anyways okay for a draft PR.
+impl<T: CheckedAdd + CheckedMul + CheckedSub + Bounded> DefensiveSaturating for T {
+	fn defensive_saturating_add(self, other: Self) -> Self {
+		self.checked_add(&other).defensive_unwrap_or_else(Bounded::max_value)
```

Comment:
`CheckedAdd` is also implemented for signed integers, so this should rather return the result of `saturating_add`. Same below for `add` and `mul`.  
Or was the idea different from what I assume? You could also allow only unsigned values.

Reply:
So I can bound the input by the `Unsigned` trait that we have in sp-arithmetic as well, but I like the idea of just piping the results to `Saturating` trait actually, rather than re-creating it.

Reply:
Okay you changed it to return `saturating_*`, that fixes my concerns.